### ISSUE TITLE
cc65 should conflict of extern with static be an error?

### ISSUE THREAD

**Issue by bbbradsmith** [cc65/cc65#2111]
This example:
```
extern void func();
static void func(void) {}
```
Compiles only with a warning:
```
Warning: Static declaration of 'func' follows non-static declaration
```
This is an error with gcc or clang at compile time.  MSVC doesn't warn at compile time, but at link time it will error with an unresolved extern.

For exploration: [Example in Godbolt](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,selection:(endColumn:1,endLineNumber:3,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:3,startColumn:1,startLineNumber:1),source:'extern+void+func()%3B%0Astatic+void+func(void)+%7B%7D%0A'),l:'5',n:'0',o:'C+source+%231',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:cc65_trunk,deviceViewOpen:'1',filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,libs:!(),options:'',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+6502+cc65+trunk+(Editor+%231)',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'x86-64+gcc+13.1',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+6502+cc65+trunk+(Compiler+%231)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

Shouldn't we want an error here? My own expectation was error, since these two things are semantically incompatible. I'm not sure whether the C spec dictates how this should be handled.

**Comment by kugelfuhr** [comment]
The code exploits undefined behavior, so all compilers are correct.

ISO/IEC 9899:1999 (E), 6.2.2p7 "Linkages of identifiers": "If, within a translation unit, the same identifier appears with both internal and external linkage, the behavior is undefined."

**Comment by mrdudz** [comment]
Don't later standards actually make it an error?

Whatever the answer is, i think i'd prefer an error over a warning too :)

**Comment by bbbradsmith** [comment]
Converting the caught warning to error should be easy, though I should note that there is not an existing warning if declared extern after the static definition. This has no warning:
```
static void func(void) {}
exern void func(void);
```
I can't find any compiler that warns or errors for this, even with -Wall, and I'm a little curious why. I can see that the extern declaration can be ignored, since if this translation unit needs func it's got one it can find without extern.

My intuition is that it's indicating an ambiguity that's worth a warning, but I suspect there's a reason why it appears to be purposefully ignored.

**Comment by mrdudz** [comment]
I wouldn't mind a warning in the second case either :)

**Comment by bbbradsmith** [comment]
I think #2112 should cover both of those wants. Warning becomes error. New warning for the case that discards the extern.

**Comment by bbbradsmith** [comment]
Previous discussion #191 had a vote that it be the warning rather than an error, and no warning had been suggested for the allowed case.

**Comment by mrdudz** [comment]
closed via #2112

**Comment by acqn** [comment]
> Edit: looking at that cited area of the spec 6.2.2.4:
> 
> > For an identifier declared with the storage-class specifier extern in a scope in which prior declaration of that identifier is visible, if the prior declaration specifies internal or external linkage, the linkage of the identifier at the later declaration is the same as the linkage specified at the prior declaration. If no prior declaration is visible, or if the prior declaration specifies no linkage, then the identifier has external linkage.
> 
> So, it is _defined_ to discard "extern" if there is a previous conflicting declaration. I can't think of a purpose for it, but I guess this is explicitly permitted by the standard in this way.

I think the purpose is to allow to legimately redeclare an identifier with internal or external linkage to pull it in the current scope in a universal way. How can that be useful? Maybe with an example:

```c
static int i;

void foo()
{
    extern int i;  // You cannot redeclare that one "static int i" in any other way!
    typedef int i; // Caught typo (intended to be something else) thanks to the "useless extern declaration"!
    /* ... */
}
```
Well, maybe it is not so useful.
